### ISSUE TITLE
core: be able to mark a stack as "absent"

### ISSUE THREAD

**Issue by diranged** [aws/aws-cdk#27444]
### Describe the feature

We'd like to see a `state` property added to the `Stack` construct, allowing us to destroy stacks via code commits.

### Use Case

We manage thousands of CloudFormation stacks today with a tool we wrote named [Kingpin](https://kingpin.readthedocs.io/en/latest/). While the tool is long in the tooth and definitely not as good as CDK, it has a killer feature for us ... the ability to mark a stack as `present` or `absent`. When you manage thousands of stacks with IAC, it's pretty critical that you are able to both create, update, _and destroy_ stacks all via code commits.

I'd like to see the `Stack` construct get a `state` flag that allows us to define that the stack should be `absent`... that way we can programmatically delete stacks without asking administrators to go through dozens of accounts and regions to delete things.

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.100.0

### Environment details (OS name and version, etc.)

N/A

**Comment by peterwoodworth** [comment]
Thanks for the request,

I'm not sure what would be the benefit to this, as opposed to commenting out a stack in code?

**Comment by diranged** [comment]
> Thanks for the request,
> 
> 
> 
> I'm not sure what would be the benefit to this, as opposed to commenting out a stack in code?

Commenting out a stack doesn't destroy it - it just stops deploying to it. I want to programatically destroy the stack

**Comment by peterwoodworth** [comment]
Right... silly me

**Comment by msambol** [comment]
@mrgrain I can take this with some guidance on allowed values for `state`, assuming this is something the team wants to support.

**Comment by mrgrain** [comment]
> When you manage thousands of stacks with IAC, it's pretty critical that you are able to both create, update, and destroy stacks all via code commits.

@diranged Could you describe your use-case further? How does the current approach of `cdk deploy` and `cdk destroy` not suit you? Would [programmatic CLI access](https://docs.aws.amazon.com/cdk/api/v2/docs/cli-lib-alpha-readme.html) help you?


The AWS CDK currently has a very intentional split between `cdk deploy` and `cdk destroy`. This way `cdk deploy` only every creates, and `cdk destroy` only ever destroys stuff. And if you don't run `cdk destroy` you know that nothing gets destroyed. If we started destroying stacks during `deploy` that would be quite the radical departure.

--- 

@msambol Will need some more infos from OP to figure out what exactly is needed before we can even start thinking about the how.

**Comment by diranged** [comment]
@mrgrain,
  In our environments, we run _thousands_ of cloudformation stacks. Most of these stacks come from a few repositories that deploy code all across our accounts and regions - often one template is deployed 50-100 times. When we want to delete a stack, we do not want any human involved in the process from the execution standpoint ... we want our teams to push a PR that says "I'd like to delete XYZ stack" (aka: `state: absent` is a flag we use in our [kingpin](https://kingpin.readthedocs.io/en/latest/kingpin.actors.aws.html#kingpin-actors-aws-cloudformation) tool quite regularly). The `cdk deploy` and `cdk destroy` commands are only used in our org by privileged developers who are doing early stage template testing. Beyond that, we only ever want a CI system to execute changes.

**Comment by mrgrain** [comment]
Makes sense. This sounds a lot like NestedStacks but pulling it a "level up" so that the CDK app is the top level. And also like you would prefer a single `cdk apply` command that takes a local state and makes reality just like it (i.e. creating and destroying stacks).

**Comment by mrgrain** [comment]
If you are interested in a workaround, something like that is fairly easy to implement by manipulating the construct tree:

```ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';

interface MagicStackProps extends cdk.StackProps {
  absent?: boolean;
}

class MagicStack extends cdk.Stack {
  public readonly absent: boolean;
  constructor(scope: Construct, id: string, props?: MagicStackProps) {
    super(scope, id, props);
    this.absent = props?.absent ?? false;
    if ((!process.env.DESTROY && this.absent) || (process.env.DESTROY && !this.absent)) {
      scope.node.tryRemoveChild(id);
    }
  }
}

const app = new cdk.App();

new MagicStack(app, 'AbsentStack', {
  absent: true
});
new MagicStack(app, 'PresentStack', {});
```

And then call CDK commands in your automation like this:

```console
$ cdk list
PresentStack
```

```console
$ DESTROY=1 cdk list
AbsentStack
```

**Comment by diranged** [comment]
We haven't tried the custom construct there - but we still consider this a big improvement to the codebase... we'd stil like to see this implemented natively.

**Comment by tmokmss** [comment]
TIL that `cdk deploy` actually calls `cdk destroy` internally when there is no resource in a stack.

https://github.com/aws/aws-cdk/blob/a8ad62c02ab83017af20e4b348ca301ead9aa0d4/packages/aws-cdk/lib/cdk-toolkit.ts#L346-L362

So you can delete a stack programatically by not defining any resources in the stack. Note that you need to pass `--version-reporting false` option to remove `CDK::Metadata` resource.

```ts
export class SomeStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: SomeStackProps) {
    super(scope, id, props);
    // when you want to delete this stack, do not define any resource to the stack.
    if (props.delete)  return;

    // other resources will be excluded from the template
    new Bucket(this, 'Bucket1', {
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });
  }
}
```

then run `cdk deploy --version-reporting false`, and your stack will be deleted.
